include/hipSYCL/runtime/instrumentation.hpp

```diff
@@ -0,0 +1,193 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef HIPSYCL_INSTRUMENTATION_HPP
+#define HIPSYCL_INSTRUMENTATION_HPP
+
+#include <cassert>
+#include <cstdint>
+#include <chrono>
+#include <future>
+#include <vector>
+#include <typeindex>
+#include <type_traits>
+
+#include "signal_channel.hpp"
+
+namespace hipsycl {
+namespace rt {
+
+// SYCL defines timestamps as uint64_t with nanosecond resolution
+class profiler_clock {
+public:
+  using rep = uint64_t;
+  using period = std::nano;
+  using duration = std::chrono::duration<rep, period>;
+  using time_point = std::chrono::time_point<profiler_clock>;
+
+  constexpr static bool is_steady = true;
+
+  static time_point now() {
+    return time_point{
+        duration{std::chrono::steady_clock::now().time_since_epoch()}};
+  }
+};
+
+class instrumentation {
+public:
+  /// Waits until an instrumentation has its result available
+  /// This does not need to be called manually, as the instrumentation_set
+  /// will call it automatically if an instrumentation result is requested.
+  virtual void wait() const = 0;
+  virtual ~instrumentation() = default;
+};
+
+template<typename T>
+inline constexpr bool is_instrumentation_v
+    = std::is_convertible_v<std::remove_cv_t<T> *, instrumentation *>;
+
+class instrumentation_set {
+public:
+  /// Wait for the given instrumentation to make results available. 
+  /// Returns nullptr if the given instrumentation was not set up.
+  template<typename Instr> const std::shared_ptr<Instr> get() const {
+    // First wait until all instrumentations have been set up
+    if(!_is_registration_complete){
+      _registration_complete_signal.wait();
+      _is_registration_complete = true;
+    }
+    
+    std::shared_ptr<Instr> i = nullptr;
+
+    for(const auto& current : _instrs) {
+      if(current.first == typeid(Instr)) {
+        i = std::static_pointer_cast<Instr>(current.second);
+      }
+    }
+
+    if(!i)
+      return nullptr;
+    // Wait for the target instrumentation to return results
+    i->wait();
+
+    return i;
+  }
+
+  template<typename Instr>
+  void add_instrumentation(std::shared_ptr<Instr> instr) {
+    assert(!_is_registration_complete);
+    for(const auto& i : _instrs) {
+      if(i.first == typeid(Instr)) {
+        assert(false && "Instrumentation already exists!");
+      }
+    }
+    std::type_index idx = typeid(Instr);
+    _instrs.push_back(std::make_pair(idx, instr));
+  }
+
+  // This will be called by the scheduler after node submission.
+  // After calling, no additional instrumentations can be added anymore.
+  void mark_set_complete() {
+    _is_registration_complete = true;
+    _registration_complete_signal.signal();
+  }
+
+private:
+  std::vector<std::pair<std::type_index, std::shared_ptr<instrumentation>>>
+      _instrs;
+
+  mutable signal_channel _registration_complete_signal;
+  mutable std::atomic<bool> _is_registration_complete = false;
```

Comment:
The idea here is to cache whether the registration is complete (a similar pattern exists in the dag node implementation), but I guess it could be argued that this is a case of premature optimization.

Reply:
I would expect futures to be implemented with atomics where possible anyway.